Source/WebDriver/WebSocketServer.h

```diff
@@ -0,0 +1,146 @@
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "CommandResult.h"
+#include "HTTPServer.h"
+#include <map>
+#include <optional>
+#include <vector>
+#include <wtf/JSONValues.h>
+#include <wtf/Vector.h>
+#include <wtf/WeakPtr.h>
+#include <wtf/text/WTFString.h>
+
+#if USE(SOUP)
+#include <wtf/glib/GRefPtr.h>
+typedef struct _SoupServer SoupServer;
+typedef struct _SoupWebsocketConnection SoupWebsocketConnection;
+#endif
+
+namespace WebDriver {
+
+class Session;
+class WebDriverService;
+
+struct Command {
+    // Defined as js-uint, but in practice opaque for the remote end
+    String id;
+    String method;
+    RefPtr<JSON::Object> parameters;
+
+    static std::optional<Command> fromData(const char* data, size_t dataLength);
+};
+
+// https://w3c.github.io/webdriver-bidi/#websocket-listener
+// A WebSocket listener is a network endpoint that is able to accept incoming WebSocket connections.
+// A WebSocket listener has a host, a port, a secure flag, and a list of WebSocket resources.
+struct WebSocketListener : public RefCounted<WebSocketListener> {
+    String host;
+    unsigned port;
+    bool secure;
+    std::vector<String> resources;
+
+    static RefPtr<WebSocketListener> create(const String& host, unsigned port, bool secure, const std::vector<String>& resources = { })
+    {
+        return adoptRef(*new WebSocketListener(host, port, secure, resources));
+    }
+
+private:
+    // Constructor is private to ensure it can only be called from within the class, specifically from the create method.
+    WebSocketListener(const String& host, unsigned port, bool secure, const std::vector<String>& resources)
+        : host(host), port(port), secure(secure), resources(resources) { }
+};
+
+
+class WebSocketMessageHandler {
+public:
+
+#if USE(SOUP)
+    typedef GRefPtr<SoupWebsocketConnection> Connection;
+#endif
+
+    struct Message {
+        // Optional connection, as the message might be a reply to a command not yet associated to a connection
```

Comment:
This is a little bit unclear. Are you referring to commands like 'status' which are not tied to a session?

Reply:
Updated the comment for clarity. It's related to `Message` objects created while we don't have the `Connection` object available, which gets assigned in the return path of the message pipeline.